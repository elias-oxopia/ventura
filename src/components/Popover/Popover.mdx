---
name: Dropdown menu
menu: Components
route: /dropdown-menu
---

import { Playground, Props } from 'docz';
import { Popover, Menu, Button } from '../../index.ts';

# Popover

A Popover is used to display extra information

## Best practices

- Check

## Examples

### Basic usage

<Playground>
  {() => {
    const [isVisible, setIsVisible] = React.useState(false);
    const popoverContent = (<p>Hello</p>);
    const onClose = () => {
      setIsVisible(false);
    }
    return (
      <>
        <Popover isVisible={isVisible} onClose={setIsVisible} content={popoverContent}>
          <Button onClick={() => setIsVisible(!isVisible)}>Open basic modal</Button>
        </ Popover>
        <br />
        <br />
        <br />
      </>
    );
  }}

</Playground>

## API

<Props of={Popover} />
