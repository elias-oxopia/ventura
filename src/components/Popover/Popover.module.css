.menu {
  border-radius: var(--border-radius-small);
  background-color: var(--color-secondary);
  box-shadow: 0px 1px 5px rgba(0, 0, 0, 0.1);
}
