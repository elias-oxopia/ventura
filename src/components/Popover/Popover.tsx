/* eslint-disable react/prop-types */
/* eslint-disable react/jsx-props-no-spreading */
import React, { useRef, useState } from 'react';
import { usePopper } from 'react-popper';
import { Placement } from '@popperjs/core';

import styles from './Popover.module.css';

interface Props {
  content?: React.ReactNode;
  children?: React.ReactNode;
  isVisible?: boolean;
  placement?: Placement;
}

const Popover: React.FC<Props> = ({
  content,
  children,
  isVisible = false,
  placement = 'bottom-start',
}) => {
  const divRef = useRef<HTMLDivElement>(null);
  const menuRef = useRef<HTMLDivElement>(null);

  const [arrowRef, setArrowRef] = useState<HTMLDivElement | null>(null);

  const { styles: popperStyles, attributes } = usePopper(divRef.current, menuRef.current, {
    modifiers: [
      {
        name: 'offset',
        options: {
          offset: [0, 10],
        },
      },
      {
        name: 'arrow',
        options: {
          element: arrowRef,
        },
      },
    ],
    placement,
  });

  return (
    <>
      <div ref={divRef}>{children}</div>

      {isVisible && (
        <div
          ref={menuRef}
          style={popperStyles.popper}
          {...attributes.popper}
          className={styles.menu}
        >
          <div ref={setArrowRef} style={popperStyles.arrow} className="arrow" />
          {content}
        </div>
      )}
    </>
  );
};

export default Popover;
