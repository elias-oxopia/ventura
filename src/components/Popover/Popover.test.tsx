import React from 'react';
import { fireEvent, render, waitFor } from '@testing-library/react';

import { Button, GitHub, Menu } from '../../index';
import Popover from './Popover';

describe('Popover', () => {
  test('default snapshot', async () => {
    const content = (
      <Menu>
        <Menu.Item name="default">Option 1</Menu.Item>
        <Menu.Item prefixIcon={<GitHub />}>Option 2</Menu.Item>
      </Menu>
    );

    const component = (
      <Popover isVisible onClose={() => {}} content={content}>
        <Button>Click me!</Button>
      </Popover>
    );
    const { asFragment, getByTestId } = render(component);

    await waitFor(() => getByTestId('menu-item-default'));
    expect(asFragment()).toMatchSnapshot();
  });

  test('click outside should trigger onClose', async () => {
    const onCloseFn = jest.fn();
    const content = (
      <Menu>
        <Menu.Item name="default">Option 1</Menu.Item>
        <Menu.Item prefixIcon={<GitHub />}>Option 2</Menu.Item>
      </Menu>
    );

    const component = (
      <>
        <Popover isVisible onClose={onCloseFn} content={content}>
          <Button>Click me!</Button>
        </Popover>
        <Button name="click">Random button!</Button>
      </>
    );
    const { getByTestId } = render(component);

    await waitFor(() => getByTestId('menu-item-default'));

    const outsideButton = getByTestId('button-click');
    fireEvent.click(outsideButton);

    expect(onCloseFn).toHaveBeenCalledTimes(1);
  });
});
