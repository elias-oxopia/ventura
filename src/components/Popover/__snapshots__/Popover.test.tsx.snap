// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popover default snapshot 1`] = `
<DocumentFragment>
  <div
    class="triggerContainer"
  >
    <button
      class="ventura button typePrimary"
      type="button"
    >
      <span
        class="label"
      >
        Click me!
      </span>
    </button>
  </div>
  <div
    class="contentContainer"
    data-popper-escaped="true"
    data-popper-placement="bottom-start"
    data-popper-reference-hidden="true"
    style="position: absolute; left: 0px; top: 0px; transform: translate(0px, 10px);"
  >
    <div
      class="arrow"
      style="position: absolute; left: 0px; transform: translate(0px, 0px);"
    />
    <div
      class="ventura menu"
    >
      <button
        class="ventura menuItem"
        data-testid="menu-item-default"
        name="default"
        type="button"
      >
        <span
          class="label"
        >
          Option 1
        </span>
      </button>
      <button
        class="ventura menuItem"
        type="button"
      >
        <svg
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M9 19c-5 1.5-5-2.5-7-3m14 6v-3.87a3.37 3.37 0 0 0-.94-2.61c3.14-.35 6.44-1.54 6.44-7A5.44 5.44 0 0 0 20 4.77 5.07 5.07 0 0 0 19.91 1S18.73.65 16 2.48a13.38 13.38 0 0 0-7 0C6.27.65 5.09 1 5.09 1A5.07 5.07 0 0 0 5 4.77a5.44 5.44 0 0 0-1.5 3.78c0 5.42 3.3 6.61 6.44 7A3.37 3.37 0 0 0 9 18.13V22"
          />
        </svg>
        <span
          class="label labelWithPrefixIcon"
        >
          Option 2
        </span>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
